### PR TITLE
docs: Clarify behaviour of multiple matchers of the same type

### PR DESCRIPTION
**Pull request by francislavoie**
Followup to https://github.com/caddyserver/caddy/pull/3832 and https://github.com/caddyserver/caddy/pull/3839 to clarify how each matchers are merged, if possible.

I'm not perfectly satisfied with the wording, so feel free to suggest revisions 👍

### CODE REVIEW

File: src/docs/markdown/caddyfile/matchers.md [hunk 1]
```diff
@@ -139,7 +139,7 @@ Like directives, named matcher definitions must go inside the site blocks that u
 
 A named matcher definition constitutes a _matcher set_. Matchers in a set are AND'ed together; i.e. all must match. For example, if you have both a `header` and `path` matcher in the set, both must match.
 
-For most matchers that accept multiple values, those values are OR'ed; i.e. one must match in order for the matcher to match.
+For some matchers that accept multiple values, those values are OR'ed; i.e. one must match in order for the matcher to match. See the docs for each matcher for details on how they are merged.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
Multiple matchers are combined using boolean algebra (AND/OR), as described by each section below.
```

File: src/docs/markdown/caddyfile/matchers.md [hunk 2]
```diff
@@ -249,14 +249,24 @@ By request header fields.
 	- If enclosed by `*`, it performs a fast substring match.
 	- Otherwise, it is a fast exact match.
 
+If used multiple times in the same named matcher, matchers with the same header field will be merged, which will OR the values of each.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]: I think this is a wee bit confusing... how about:

```suggestion
Different header fields within the same set are AND-ed. Multiple values per field are OR-ed.
```

File: src/docs/markdown/caddyfile/matchers.md [hunk 3]
```diff
@@ -267,6 +277,8 @@ header_regexp [<name>] <field> <regexp>
 
 Like `header`, but supports regular expressions. Capture groups can be accessed via placeholder like `{http.regexp.name.capture_group}` where `name` is the name of the regular expression (optional, but recommended) and `capture_group` is either the name or number of the capture group in the expression. Capture group `0` is the full regexp match, `1` is the first capture group, `2` is the second capture group, and so on.
 
+There can only be multiple `header_regexp` matchers per named matcher, but only one per field. Each field will be AND'ed.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]: The rest of these also read a little awkward to me, what if we try this:

```suggestion
Only one regular expression is supported per header field. Multiple different fields will be AND-ed.
```

File: src/docs/markdown/caddyfile/matchers.md [hunk 4]
```diff
@@ -285,6 +297,8 @@ host <hosts...>
 
 Matches request by the `Host` header field of the request. It is not common to use this in the Caddyfile, since most site blocks already indicate hosts in the address of the site. This matcher is mostly used in site blocks that don't define specific hostnames.
 
+There can be multiple `host` matchers per named matcher, merging them. The hosts will be OR'ed.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
Multiple `host` matchers will be OR'ed together.
```

File: src/docs/markdown/caddyfile/matchers.md [hunk 5]
```diff
@@ -301,6 +315,8 @@ method <verbs...>
 
 By the method (verb) of the HTTP request. Verbs should be uppercase, like `POST`. Can match one or many methods.
 
+There can be multiple `method` matchers per named matcher, merging them. The verbs will be OR'ed.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
Multiple `method` matchers will be OR'ed together.
```

File: src/docs/markdown/caddyfile/matchers.md [hunk 6]
```diff
@@ -380,6 +396,8 @@ By request path, meaning the path component of the request's URI. Path matches a
 - On both sides, for a substring match (`*/contains/*`)
 - In the middle, for a globular match (`/accounts/*/info`)
 
+There can be multiple `path` matchers per named matcher, merging them. The paths will be OR'ed.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
Multiple `path` matchers will be OR'ed together.
```

File: src/docs/markdown/caddyfile/matchers.md [hunk 7]
```diff
@@ -416,7 +438,9 @@ By request protocol.
 query <key>=<val>...
 ```
 
-By query string parameters. Should be a sequence of `key=value` pairs. Keys are matched exactly, case-sensitively. Values are matched exactly, but also support `*` to match any value.
+By query string parameters. Should be a sequence of `key=value` pair. Keys are matched exactly, case-sensitively. Values can contain placeholders. Values are matched exactly, but also support `*` to match any value.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
By query string parameters. Should be a sequence of `key=value` pairs. Keys are matched exactly, case-sensitively. Values can contain placeholders. Values are matched exactly, but also support `*` to match any value.
```

File: src/docs/markdown/caddyfile/matchers.md [hunk 8]
```diff
@@ -416,7 +438,9 @@ By request protocol.
...
+
+There can be multiple `query` matchers per named matcher, and pairs with the same keys will be merged together, which will OR the values of each.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
There can be multiple `query` matchers per named matcher, and pairs with the same keys will be OR'ed together.
```

File: src/docs/markdown/caddyfile/matchers.md [hunk 9]
```diff
@@ -436,6 +460,8 @@ remote_ip <ranges...>
 
 By remote (client) IP address. If `X-Forwarded-For` is passed in request headers, this will be used for remote IP address. Accepts exact IPs or CIDR ranges.
 
+There can be multiple `remote_ip` matchers per named matcher, merging them. The addresses will be OR'ed.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
Multiple `remote_ip` matchers will be OR'ed together.
```

**Review by mholt — CHANGES_REQUESTED** [review]
Thanks! I did find the phrasing a little confusing/awkward on a lot of these, so I made some suggestions to simplify them. Let me know what you think.

File: src/docs/markdown/caddyfile/matchers.md (hunk 1 above)
Comment (francislavoie) [reply src/docs/markdown/caddyfile/matchers.md]: I think 
```suggestion
Multiple matchers of the same type are combined using boolean algebra (AND/OR), as described by each section below.
```

i.e. + "of the same type"

Comment (mholt) [reply src/docs/markdown/caddyfile/matchers.md]: Even different matchers are combined using AND logic though.

Comment (francislavoie) [reply src/docs/markdown/caddyfile/matchers.md]: True. I think we should separate those ideas then. It should be clear that matchers of a different type will always be ANDed but matchers of the same type may be AND or OR depending.

Comment (francislavoie) [reply src/docs/markdown/caddyfile/matchers.md]: Updated, went with:

```
Multiple matchers of the same type (e.g. multiple `path` matchers in the same set) may be combined using boolean algebra (AND/OR), as described in their respective sections below.
```

I think that's probably okay.

File: src/docs/markdown/caddyfile/matchers.md [hunk 10]
```diff
@@ -267,6 +277,8 @@ header_regexp [<name>] <field> <regexp>
 
 Like `header`, but supports regular expressions. Capture groups can be accessed via placeholder like `{http.regexp.name.capture_group}` where `name` is the name of the regular expression (optional, but recommended) and `capture_group` is either the name or number of the capture group in the expression. Capture group `0` is the full regexp match, `1` is the first capture group, `2` is the second capture group, and so on.
 
+Only one regular expression is supported per header field. Multiple different fields will be AND-ed.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
Only one regular expression is supported per header field. Multiple different fields will be AND'ed.
```

File: src/docs/markdown/caddyfile/matchers.md [hunk 11]
```diff
@@ -139,7 +139,7 @@ Like directives, named matcher definitions must go inside the site blocks that u
 
 A named matcher definition constitutes a _matcher set_. Matchers in a set are AND'ed together; i.e. all must match. For example, if you have both a `header` and `path` matcher in the set, both must match.
 
-For most matchers that accept multiple values, those values are OR'ed; i.e. one must match in order for the matcher to match.
+Multiple matchers of the same type (e.g. multiple `path` matchers in the same set) may be combined using boolean algebra (AND/OR), as described in their respective sections below.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
Multiple matchers of the same type may be combined (e.g. multiple `path` matchers in the same set) using boolean algebra (AND/OR), as described in their respective sections below.
```

Just to make it clearer that we're not talking about combining multiple sets, but multiple matchers within a set.

File: src/docs/markdown/caddyfile/matchers.md [hunk 12]
```diff
@@ -249,14 +249,24 @@ By request header fields.
 	- If enclosed by `*`, it performs a fast substring match.
 	- Otherwise, it is a fast exact match.
 
+Different header fields within the same set are AND-ed. Multiple values per field are OR-ed.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/matchers.md]:
```suggestion
Different header fields within the same set are AND-ed. Multiple values per field are OR'ed.
```

**Review by mholt — CHANGES_REQUESTED** [review]
A couple minor changes for consistency and clarity. Almost there!

**Review by mholt — APPROVED** [review]

**Pull request merged**